### PR TITLE
[constraint] adds PositionConstraint with example

### PR DESCRIPTION
**Pull request by EulalieCoevoet**
This PR adds two classes:
- PositionConstraint: this component allows to impose a displacement to a set of points, using a constraint approach with Lagrange's multipliers. See example `docs/sofapython3/examples/component/constraint/PositionConstraint/PositionConstraint.py`. 
- PositionModel: shared implementation of PositionConstraint and PositionEffector (for inverse resolution).

### CODE REVIEW

File: src/SoftRobots/component/constraint/model/PositionModel.inl [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        return;
+
+    vector<Coord> points;
+    ReadAccessor<Data<VecCoord> > positions = m_state->readPositions();
+    ReadAccessor<Data<type::vector<sofa::Index>> > indices = d_indices;
+    for (unsigned int i=0; i<indices.size(); i++)
+    {
+        points.push_back(positions[indices[i]]);
+    }
+    drawPoints(vparams, points, 10.0f, RGBAColor(0.,1.,0.,1.));
```
Comment (alxbilger) [inline src/SoftRobots/component/constraint/model/PositionModel.inl:298]:
```suggestion
    drawPoints(vparams, points, 10.0f, RGBAColor::green());
```

File: src/SoftRobots/component/constraint/model/PositionModel.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+template<>
+void PositionModel<Vec3Types>::drawPoints(const VisualParams* vparams, const std::vector<Coord> &points, float size,  const  RGBAColor& color) {
+    vparams->drawTool()->drawPoints(points, size, color);
+}
+
+template<>
+void PositionModel<Vec2Types>::drawPoints(const VisualParams* vparams, const std::vector<Coord> &points, float size,  const  RGBAColor& color) {
+    vector<Vec3> pointsVec3;
```
Comment (alxbilger) [inline src/SoftRobots/component/constraint/model/PositionModel.cpp:68]:
```suggestion
    vector<Vec3> pointsVec3;
    pointsVec3.reserve(points.size());
```

File: src/SoftRobots/component/constraint/model/PositionModel.cpp
Comment (alxbilger) [inline src/SoftRobots/component/constraint/model/PositionModel.cpp:1]: I think you can factorize the `drawPoints` method using `DataTypes::getCPos`

File: src/SoftRobots/component/constraint/PositionConstraint.inl [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+    , d_value(initData(&d_value, "value",
+                                "Displacement or force to impose.\n"))
+
+    , d_valueIndex(initData(&d_valueIndex, (unsigned int) 0, "valueIndex",
+                                  "Index of the value (in InputValue vector) that we want to impose \n"
+                                  "If unspecified the default value is {0}"))
+
+    , d_valueType(initData(&d_valueType, OptionsGroup(2,"displacement","force"), "valueType",
+                                          "displacement = the contstraint will impose the displacement provided in data value[valueIndex] \n"
```
Comment (alxbilger) [inline src/SoftRobots/component/constraint/PositionConstraint.inl:83]:
```suggestion
                                          "displacement = the constraint will impose the displacement provided in data value[valueIndex] \n"
```

File: src/SoftRobots/component/constraint/PositionConstraint.inl [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                                          "force = the contstraint will impose the force provided in data value[valueIndex] \n"
```
Comment (alxbilger) [inline src/SoftRobots/component/constraint/PositionConstraint.inl:84]:
```suggestion
                                          "force = the constraint will impose the force provided in data value[valueIndex] \n"
```

File: src/SoftRobots/component/constraint/PositionConstraint.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+.add< PositionConstraint<Vec2Types> >()
+.add< PositionConstraint<Rigid3Types> >()
+
+;
+////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// Force template specialization for the most common sofa type.
+// This goes with the extern template declaration in the .h. Declaring extern template
+// avoid the code generation of the template for each compilation unit.
+// see: http://www.stroustrup.com/C++11FAQ.html#extern-templates
```
Comment (alxbilger) [inline src/SoftRobots/component/constraint/PositionConstraint.cpp:133]:
```suggestion
```

**Review by alxbilger — CHANGES_REQUESTED** [review]

File: src/SoftRobots/component/constraint/model/PositionModel.inl [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+    cIndex += index;
+    cMatrix.endEdit();
+
+    m_nbLines = cIndex - m_constraintId;
+}
+
+
+template<class DataTypes>
+void PositionModel<DataTypes>::storeResults(vector<double> &delta)
```
[the remark this answers is not in the document]
Comment (EulalieCoevoet) [reply src/SoftRobots/component/constraint/model/PositionModel.inl:237]: No, here the idea is to allow the component to change the value if needed.

**Review by alxbilger — CHANGES_REQUESTED** [review]
Only minor changes required, otherwise go to go

**Comment by EulalieCoevoet** [comment]
Thanks @alxbilger for your review. I think we can merge now.

**Pull request merged**